### PR TITLE
Clee iteration one refactor

### PR DESCRIPTION
**Pull request by CaptainSpectacular**
Just refactoring what we had.

### CODE REVIEW

File: test/sales_analyst_test.rb [hunk 1]
```diff
@@ -17,14 +17,83 @@ def test_existence
...
+    assert_equal 3.26, sa.average_items_per_merchant_standard_deviation
+  end
+
+  def test_merchants_with_high_item_count_method
+    skip
+    sa = SalesAnalyst.new(@se)
+
+    assert_instance_of Array, sa.merchants_with_high_item_count
+    assert_instance_of Merchant, sa.merchants_with_high_item_count[0]
+    assert_equal 114, sa.merchants_with_high_item_count.size # tough to calculate on our own...seems right, but we need to check this value manually (with fixtures?)
```
[the remark this answers is not in the document]
Comment (CaptainSpectacular) [reply test/sales_analyst_test.rb:37]: Will be fixed when we go back to make this test work. comment will disappear.

File: test/item_test.rb [hunk 2]
```diff
@@ -3,14 +3,25 @@
 require 'bigdecimal'
 
 class ItemTest < MiniTest::Test
+  def setup
+    @attributes = { id: 1,
+                    name: 'Pencil',
+                    description: 'You can use it to write things',
+                    unit_price: BigDecimal.new(10.99, 4),
+                    merchant_id:1,
```
[the remark this answers is not in the document]
Comment (CaptainSpectacular) [reply test/item_test.rb]: fixed.

File: lib/sales_analyst.rb [hunk 3]
```diff
@@ -1,16 +1,44 @@
...
+require 'bigdecimal'
 # This is the SalesAnalyst class
 class SalesAnalyst
-  include Arithmatic
+  include Arithmetic
+  attr_reader :standard_deviation_avg_item_price
   def initialize(sales_engine)
     @sales_engine = sales_engine
+    @standard_deviation_items = average_items_per_merchant_standard_deviation
+    # @standard_deviation_avg_item_price = average_average_item_prices_per_merchant_standard_deviation
```
[the remark this answers is not in the document]
Comment (CaptainSpectacular) [reply lib/sales_analyst.rb:10]: Will be fixed when we go back to make this test work. comment will disappear.

File: lib/arithmetic.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    BigDecimal(sales_engine.items.find_all_by_merchant_id(id).size)
+  end
+
+  def sum_unit_prices_for_merchant(sales_engine, id)
+    sales_engine.items.find_all_by_merchant_id(id).reduce(BigDecimal.new(0)) do |sum, item|
+      sum += item.unit_price_to_dollars
+    end
+  end
+
+  # moved core formula here because of the need to carry out division over two lines
```
[the remark this answers is not in the document]
Comment (CaptainSpectacular) [reply lib/arithmetic.rb:33]: Will be fixed when we go back to make this test work. comment will disappear.

File: lib/arithmetic.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    Math.sqrt(variance_items_per_merchant(sales_engine))
+  end
+
+  def total_units_for_merchant(sales_engine, id)
+    BigDecimal(sales_engine.items.find_all_by_merchant_id(id).size)
+  end
+
+  def sum_unit_prices_for_merchant(sales_engine, id)
+    sales_engine.items.find_all_by_merchant_id(id).reduce(BigDecimal.new(0)) do |sum, item|
+      sum += item.unit_price_to_dollars
```
[the remark this answers is not in the document]
Comment (CaptainSpectacular) [reply lib/arithmetic.rb:29]: will refactor at end of iteration

File: lib/arithmetic.rb [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  def std_deviation_items_per_merchant(sales_engine)
+    Math.sqrt(variance_items_per_merchant(sales_engine))
+  end
+
+  def total_units_for_merchant(sales_engine, id)
+    BigDecimal(sales_engine.items.find_all_by_merchant_id(id).size)
+  end
+
+  def sum_unit_prices_for_merchant(sales_engine, id)
+    sales_engine.items.find_all_by_merchant_id(id).reduce(BigDecimal.new(0)) do |sum, item|
```
[the remark this answers is not in the document]
Comment (CaptainSpectacular) [reply lib/arithmetic.rb:28]: will refactor in future commit.

File: lib/arithmetic.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  end
+
+  def sum_diff_means_items_per_merchant_squared(sales_engine)
+    sales_engine.merchants.all.map do |merchant|
+      (merchant.items.size - avg_items_per_merchant(sales_engine))**2
+    end.sum
+  end
+
+  def variance_items_per_merchant(sales_engine)
+    sum_diff_means_items_per_merchant_squared(sales_engine) /\
```
[the remark this answers is not in the document]
Comment (CaptainSpectacular) [reply lib/arithmetic.rb:15]: will refactor in later commit

File: lib/arithmetic.rb [hunk 8]
```diff
@@ -0,0 +1,65 @@
+require 'bigdecimal'
+# This is the Arithmatic module
+module Arithmetic
+  def avg_items_per_merchant(sales_engine)
+    BigDecimal(sales_engine.items.size) / BigDecimal(sales_engine.merchants.size)
```
[the remark this answers is not in the document]
Comment (CaptainSpectacular) [reply lib/arithmetic.rb:5]: will refactor in later commit.

File: test/sales_analyst_test.rb [hunk 9]
```diff
@@ -17,14 +17,83 @@ def test_existence
...
+    sa = SalesAnalyst.new(@se)
+
+    assert_instance_of BigDecimal, sa.sum_unit_prices_for_merchant(12_334_105)
+    assert_equal 4_997, sa.sum_unit_prices_for_merchant(12_334_105)
+  end
+
+  def test_average_item_price_for_merchant_method
+    sa = SalesAnalyst.new(@se)
+    actual_1 = sa.average_item_prices_for_merchant(12_334_105).round(2)
+    actual_2 = sa.average_item_prices_for_merchant(12_334_112).round(2)
```
[the remark this answers is not in the document]
Comment (CaptainSpectacular) [reply test/sales_analyst_test.rb:59]: conflicting infromation in terms of readability

File: test/sales_analyst_test.rb [hunk 10]
```diff
@@ -17,14 +17,83 @@ def test_existence
...
+    skip # moved this method to arithmatic
+    sa = SalesAnalyst.new(@se)
+
+    assert_instance_of BigDecimal, sa.sum_unit_prices_for_merchant(12_334_105)
+    assert_equal 4_997, sa.sum_unit_prices_for_merchant(12_334_105)
+  end
+
+  def test_average_item_price_for_merchant_method
+    sa = SalesAnalyst.new(@se)
+    actual_1 = sa.average_item_prices_for_merchant(12_334_105).round(2)
```
[the remark this answers is not in the document]
Comment (agpiermarini) [reply test/sales_analyst_test.rb:58]: conflicting infromation in terms of readability

**Pull request merged**